src/plugins/vis_type_timeline/server/series_functions/graphite.js

```diff
@@ -35,6 +35,81 @@ import _ from 'lodash';
 import fetch from 'node-fetch';
 import moment from 'moment';
 import Datasource from '../lib/classes/datasource';
+import dns from 'dns-sync';
+import IPCIDR from 'ip-cidr';
+
+const MISS_CHECKLIST_MESSAGE = `Please configure on the opensearch_dashbpards.yml file. 
+You can always enable the default allowlist configuration.`;
+
+const INVALID_URL_MESSAGE = `The Graphite URL/IP provided by you is invalid. 
+Please update your config from OpenSearch Dashboards's Advanced Setting.`;
+
+/**
+ * Resolve hostname to IP address
+ * @param {object} urlObject
+ * @returns {string} configuredIP
+ * or null if it cannot be resolve
+ * According to RFC, all IPv6 IP address needs to be in []
+ * such as [::1]
+ * So if we detect a IPv6 address, we remove brackets
+ */
+function getIpAddress(urlObject) {
+  const hostname = urlObject.hostname;
+  const configuredIP = dns.resolve(hostname);
+  if (configuredIP) {
+    return configuredIP;
+  }
+  if (hostname.startsWith('[') && hostname.endsWith(']')) {
+    return hostname.substr(1).slice(0, -1);
+  }
+  return null;
+}
+
+function isValidURL(configuredUrl, blockedIPs) {
+  // Check the format of URL, URL has be in the format as
+  // scheme://server/path/resource otherwise an TypeError
+  // would be thrown
+  let configuredUrlObject;
+  try {
+    configuredUrlObject = new URL(configuredUrl);
+  } catch (err) {
+    return false;
+  }
+
+  const ip = getIpAddress(configuredUrlObject);
+  if (!ip) {
+    return false;
+  }
+
+  // IPCIDR check if a specific IP address fall in the
+  // range of an IP address block
+  // @param {string} bl
+  // @returns {object} cidr
+  const isBlocked = blockedIPs.some((blockedIP) => new IPCIDR(blockedIP).contains(ip));
+  return !isBlocked;
+}
+
+/**
+ * Check configured url using blocklist and allowlist
+ * If allowlist is used, return false if allowlist does not contain configured url
+ * If blocklist is used, return false if blocklist contains configured url
+ * If both allowlist and blocklist are used, check blocklist first then allowlist
+ * @param {Array|string} blockedIPs
+ * @param {Array|string} allowedUrls
+ * @param {string} configuredUrls
+ * @returns {boolean} true if the configuredUrl is valid
+ */
+function isValidConfig(blockedIPs, allowedUrls, configuredUrl) {
```

Comment:
could we add unit test for this function.

btw, not required for this PR, but please consider code coverage support when run test.

Reply:
To write a unit test, I need to import isValidConfig in graphite.test.js. But change to `export funtion isValidConfig` to graphite.test.js will trigger en error:
<img width="354" alt="Screen Shot 2021-06-02 at 10 42 54 PM" src="https://user-images.githubusercontent.com/79961084/120596810-0a2e1f80-c3f9-11eb-8de2-f54b7c3774a4.png">

I will check if there could be other ways to write unit tests.

Reply:
in graphite.js, add a wrapper to export internal method without change
```
// use this
function isValidConfigInternal(...) {
  // implementation
}

// export for testability
export function isValidConfig(...) {
  isValidConfigInternal(...)
}
```

in test file
```
import {isValidConfig} from './graphite';

  it('test isValidConfig', function(){
        expect(isValidConfig(.....)).to.eql(.....);
  })
```